src/Catalyst.Core.Modules.Consensus/Deltas/DeltaHashProvider.cs

```diff
@@ -0,0 +1,126 @@
+        private readonly int _capacity;
+
+        public IObservable<Cid> DeltaHashUpdates => _deltaHashUpdatesSubject.AsObservable();
+
+        public DeltaHashProvider(IDeltaCache deltaCache,
+            ILogger logger,
+            int capacity = 10_000)
+        {
+            _deltaCache = deltaCache;
+            _logger = logger;
+            _deltaHashUpdatesSubject = new ReplaySubject<Cid>(0);
+            var comparer = ComparerBuilder.For<Timestamp>().OrderBy(u => u, descending: true);
+            _capacity = capacity;
+            _hashesByTimeDescending = new SortedList<Timestamp, Cid>(comparer)
+            {
+                Capacity = _capacity
+            };
+
+            _hashesByTimeDescending.Add(Timestamp.FromDateTime(DateTime.MinValue.ToUniversalTime()),
+                _deltaCache.GenesisHash);
+        }
+
+        /// <inheritdoc />
+        public bool TryUpdateLatestHash(Cid previousHash, Cid newHash)
+        {
+            var newAddress = newHash;
+            var previousAddress = previousHash;
+            _logger.Debug("New hash {hash} received for previous hash {previousHash}",
+                newAddress, previousAddress);
+            var foundNewDelta = _deltaCache.TryGetOrAddConfirmedDelta(newAddress, out var newDelta);
+            var foundPreviousDelta = _deltaCache.TryGetOrAddConfirmedDelta(previousAddress, out var previousDelta);
+
+            if (!foundNewDelta
+             || !foundPreviousDelta
+             || newDelta.PreviousDeltaDfsHash != previousHash.ToArray().ToByteString()
+             || previousDelta.TimeStamp >= newDelta.TimeStamp)
+            {
+                _logger.Warning("Failed to update latest hash from {previousHash} to {newHash}",
+                    previousAddress, newAddress);
+                return false;
+            }
+
+            _logger.Debug("Successfully to updated latest hash from {previousHash} to {newHash}",
+                previousAddress, newAddress);
+
+            lock (_hashesByTimeDescending)
+            {
+                _hashesByTimeDescending.Add(newDelta.TimeStamp, newHash);
+                if (_hashesByTimeDescending.Count > _capacity)
+                {
+                    _hashesByTimeDescending.RemoveAt(_capacity);
+                }
+            }
+
+            _deltaHashUpdatesSubject.OnNext(newAddress);
+
+            return true;
+        }
+
+        /// <inheritdoc />
+        public Cid GetLatestDeltaHash(DateTime? asOf = null)
+        {
+            _logger.Verbose("Trying to retrieve latest delta as of {asOf}", asOf);
+            if (!asOf.HasValue)
+            {
+                return _hashesByTimeDescending.FirstOrDefault().Value;
+            }
+
+            var timestamp = Timestamp.FromDateTime(asOf.Value);
+            var hash = _hashesByTimeDescending
+               .SkipWhile(p => p.Key > timestamp)
+               .FirstOrDefault();
+
+            //todo: do we want to start walking down
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)